### ISSUE TITLE
Endless runner rewards

### ISSUE THREAD

**Issue by AidinD** [Unity-Technologies/ml-agents#2396]
// EDIT: I found the original issue, but I am still interested in knowing below:

I am trying to make a flappy bird learner. But instead of using the pipes positions as observations I want to only use observations that the bird itself knows. Meaning, my observations are:

A lot of RayPerceptions
Its own y position (probably not needed)
Its own y velocity (is this needed?)

How should you generally think when you are making rewards for an endless runner?

When is the reward getting evaluated? Is it continously or only when done() is called? I mean, for an endless runner, it is preferable for the runner to keep going, hence done() will never be called and result in a lot of: 
`No episode was completed since last summary. Training.`
Or am I mistaken somehow?

Here is the entire AgentAction method

    public override void AgentAction(float[] vectorAction, string textAction) {
        var action = vectorAction[0];

        switch (action) {
            case 0: {
                    break;
                }
            case 1: {
                    rigidbody.velocity = Vector2.up * speed;
                    break;
                }
        }

        // Rewards
        AddReward(0.001f);
        if (collidedGround || collidedPipe) {
            SetReward(-1);
            Done();
        }
    }

**Comment by awjuliani** [comment]
Hi @AidinD 

Rewards are calculated every agent step. In the case of this game, I would add a small positive reward for staying alive, and a negative reward for death, as you have here.

**Comment by AidinD** [comment]
Hi,
Yes, I understand that they are calculated every step. But are they also evaluated by the machine every step?
Is the machine thinking, alright, this step I got positive points, I did something right, next step I should try to do something similar?
The thing that bothers me seems to be the fact that we never call done() if the machine does everything right, and I have come to believe that the done() is crucial for the evaluation and therefor I feel like if the machine happens to be good then done() never runs and then the machine will never understand that he did good?
